docs/strings.md

```diff
@@ -0,0 +1,464 @@
+	 */
+	func fastest_encoding(self: readonly & String) : StringEncoding;
+
+	/**
+	 * The standard encoding that this string can most easily convert to.  This
+	 * will typically be the encoding that the string uses internally but may
+	 * not be for custom encodings.  For all standard-library string types,
+	 * this will return the same value as `fastest_encoding`.
+	 */
+	func fastest_unicode_encoding(self: readonly & String) : UnicodeStringEncoding;
+
+	/**
+	 * The standard encoding that will use the smallest amount of space to
+	 * store the characters.  For example, most English strings will return
+	 * `StringEncodingASCII`, French will typically return
+	 * `StringEncodingUTF8`, whereas CJK language strings will likely return
+	 * `StringEncodingUTF16`.
+	 */
+	func smallest_unicode_encoding(self: readonly & String) : UnicodeStringEncoding;
+
+	/**
+	 * Access a specific character.  Returns a unicode code point for the
+	 * specific index in the string.
+	 */
+	func apply(self: readonly & String, index: U64) : Unichar;
+
+	/**
+	 * Access a specific character.
+	 */
+	func apply(self: readonly & String, index: U64) : Unichar;
+
+	/**
+	 * Takes a range in the string (expressed in Unicode code points) and
+	 * optionally a buffer.  Either the desired characters (in the specified
+	 * encoding) are copied into that buffer, or returns an immutable array
+	 * that may point to the internal storage for the string.
+	 *
+	 * If the caller first calls `fastest_unicode_encoding` and uses the result
+	 * as the generic parameter then it will return an internal buffer for
+	 * immutable strings.  This is intended to be used to build fast iterators.
+	 */
+	func copy_or_view_data[StringEncoding Enc](self: readonly & String,
+	                                           range: Range,
+	                                           buffer: Optional[Array[Enc.CharType] & mut])
+	     : Array[Enc.CharType] & (mut | imm);
+
+	/******************************************************************************
+	 * Operations on mutable strings
+	 *****************************************************************************/
+
+	/**
+	 * Updates a specific character. 
+	 */
+	func update(self: (mut | iso) & String, index: U64, c: Unichar);
+
+	/**
+	 * Replace the specified range in a string with another string.
+	 */
+	func update_range[String Str](self: (mut | iso) & String,
+	                              range: Range,
+	                              newString: Str & imm);
+}
+```
+
+For discussion, the above proposal is somewhat simplified:
+The `update` operation should take a `StringView`, a non-owning view of a substring.
+The `String` interface should also include an operation for providing a `StringView` on a specific range.
+The Verona region system makes tracking ownership of string views relatively simple:
+
+ - By default, all mutable strings are `iso`.
+ - Mutable string views are `mut`, in the same region as their owning string.
```

Comment:
Right, if they were iso they couldn't refer to the string because it would be in a different region (or, rather, they would be the only thing that could refer to the string).